libs/editor/egui_editor/src/editor.rs

```diff
@@ -58,24 +64,43 @@ impl Editor {
     pub fn ui(&mut self, ui: &mut Ui) {
         self.debug.frame_start();
 
-        self.visual_appearance.update(ui);
-        self.key_events(ui);
-        self.mouse_events(ui);
-
-        if self.text_unprocessed {
-            self.ast = Ast::parse(&self.raw);
-            self.calc_unicode_segs();
+        // update theme
+        self.appearance.current_theme =
+            if ui.visuals().dark_mode { Theme::Dark } else { Theme::Light };
+
+        // process events
+        let (text_updated, selection_updated) = if self.initialized {
+            events::process(
+                &ui.ctx().input().events,
+                &self.galleys,
+                &mut self.buffer,
+                &mut self.segs,
+                &mut self.cursor,
+                &mut self.debug,
+            )
+        } else {
+            self.segs = unicode_segs::calc(&self.buffer);
+            (true, true)
+        };
+
+        // recalculate dependent state
+        if text_updated {
+            self.ast = ast::calc(&self.buffer);
         }
-        if self.text_unprocessed || self.cursor_unprocessed {
-            self.populate_styled();
-            self.populate_layouts();
+        if text_updated || selection_updated {
+            self.styles = styles::calc(&self.ast, &self.cursor.selection_bytes(&self.segs));
```

Comment:
`_::calc` code style clarifies what data is calculated using what other data, mutably or immutably